File: src/main/java/org/bk/graphql/service/BookService.java
package org.bk.graphql.service;

import org.bk.graphql.domain.Author;
import org.bk.graphql.domain.Book;
import org.bk.graphql.domain.BookId;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public interface BookService {
    Book createBook(CreateBookCommand command);
    Book createOrUpdateBook(CreateOrUpdateBookCommand command);
    Book updateBook(UpdateBookCommand command);
    Page<Book> getBooks(GetBooksQuery query);
    Page<Book> getBooks(Pageable pageable);
    Optional<Book> getBook(ById<BookId> query);
    List<Book> getBooks(ByIds<BookId> query);
    Map<BookId, List<Author>> getAuthorsForBooks(ByIds<BookId> ids);
}

File: src/main/kotlin/org/bk/graphql/service/BookService.kt
package org.bk.graphql.service

import org.bk.graphql.domain.Author
import org.bk.graphql.domain.Book
import org.bk.graphql.domain.BookId
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import java.util.Optional

interface BookService {
    fun createBook(command: CreateBookCommand): Book
    fun createOrUpdateBook(command: CreateOrUpdateBookCommand): Book
    fun updateBook(command: UpdateBookCommand): Book
    fun getBooks(query: GetBooksQuery): Page<Book>
    fun getBooks(pageable: Pageable): Page<Book>
    fun getBook(query: ById<BookId>): Optional<Book>
    fun getBooks(query: ByIds<BookId>): List<Book>
    fun getAuthorsForBooks(ids: ByIds<BookId>): Map<BookId, List<Author>>
}